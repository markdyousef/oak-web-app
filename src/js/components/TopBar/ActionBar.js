// @flow
import React, { Component } from 'react';
import { RoundButton } from '../shared/Button';
import { NavCenter } from './styles';
import { CardUpdate, CardCreate } from './index';
import { changeUrls } from '../../utils';
import { convertToRaw } from 'draft-js';

type DefaultProps = {
    updateCard: () => void
};

type Props = {
    cardId?: string,
    card?: Object,
    collectionId?: string,
    location: {
        pathname?: string
    },
    router: {
        push?: (path: string) => void,
        replace?: (path: string) => void
    },
    params?: { cardId?: string },
    showLabels?: bool,
    updateLabels?: (field: Object) => void,
    updateCard: (key: string, value: any) => void,
    create?: (id: string, name: string, content: string, cover: string) => Promise<>,
    update?: (cardId: string, content: string, cover: string) => Promise<>,
    data?: { loading: bool, collections: Array<Object> }
};

type State = {};

export default class ActionBar extends Component<DefaultProps, Props, State> {
    static defaultProps: DefaultProps = {
        router: {},
        location: {},
        updateCard: () => {}
    }
    state: State = {}
    onShowLabels = (show?:bool = true) => {
        const { updateLabels } = this.props;
        if (!updateLabels) return;
        updateLabels({
            key: 'showLabels',
            value: show
        });
    }
    onSave = () => {
        const { create, update, updateCard, card, router: { replace } } = this.props;
        if (!card) return;

        const cardId = card.get('cardId');
        const collectionId = card.get('collectionId');
        const editorState = card.get('editorState');
        const name = card.get('name');
        const images = card.get('images');
        const newEditorState = changeUrls(editorState, images);
        const content = JSON.stringify(convertToRaw(newEditorState.getCurrentContent()));
        const cover = images[0] && images[0].id;
        // existing cards has a cardId
        updateCard('isLoading', true);
        // no card name
        // if (!name) {
        //     const message = {
        //         type: 'error',
        //         message: 'Please provide a name for your card',
        //         onClick: this.onSave
        //     }
        //     this.updateCard('message', message);
        //     return;
        // }
        if (cardId && update) {
            update(cardId, content, cover);
        } else if (create) {
            create(collectionId, name, content, cover);
        }
        if (replace) replace(`/collection/${collectionId}/card/${cardId}`)
    }
    onCreate = () => {
        const { updateCard } = this.props;
        updateCard('menu', "CREATE");
    }
    renderActionBar = () => {
        const {
            location,
            router: { push },
            collectionId,
            card,
            showLabels,
            updateCard,
            data,
            params,
            showComments,
            onShowComments
        } = this.props;
        const inCardDetail =
            location.pathname &&
            location.pathname.includes('card');

        if (!inCardDetail) {
            return(
                <RoundButton
                    onClick={() => push && push('/collection/card')}
                    text="Create Card"
                    type="secondary"
                />
            )
        }
        const isLoading = card && card.get('isLoading');
        const isSaved = card && card.get('isSaved');
        const cardId = params && params.cardId;
        //Existing Card
        if (inCardDetail && cardId) {
            // open existing card in readOnly
            const readOnly = card && card.get('readOnly');
            return (
                <CardUpdate
                    onShowLabels={this.onShowLabels}
                    showLabels={showLabels}
                    readOnly={readOnly}
                    onEdit={() => updateCard('readOnly', !readOnly)}
                    onSave={this.onSave}
                    isLoading={isLoading}
                    onShowComments={() => onShowComments(!showComments)}
                />
            );
        }
        //New Card
        if (inCardDetail && !cardId) {
            const menu = card && card.get('menu');
            return (
                <CardCreate
                    isLoading={isLoading}
                    onCreate={this.onCreate}
                    showMenu={menu === "CREATE"}
                    closeMenu={() => updateCard('menu', null)}
                    collectionId={collectionId}
                    collections={data && data.collections}
                    updateCollection={collectionId => updateCard('collectionId', collectionId)}
                    showLabels={showLabels}
                    onShowLabels={this.onShowLabels}
                    saveCard={this.onSave}
                />
            )
        }
    }
    render() {
        return (
            <NavCenter>
                {this.renderActionBar()}
            </NavCenter>
        );
    }
}
