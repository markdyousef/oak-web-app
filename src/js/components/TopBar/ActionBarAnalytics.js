// @flow
import React, { Component } from 'react';
import ReactGA from 'react-ga';


export default (ActionBar: Function) => {
    return (
        class AnalyticsWrapper extends Component {
            track = (type: string) => {
                ReactGA.event({
                    category: 'Card',
                    action: type
                });
            }
            render() {
                return (
                    <ActionBar
                        track={this.track}
                        {...this.props}
                    />
                );
            }
        }
    );
};
