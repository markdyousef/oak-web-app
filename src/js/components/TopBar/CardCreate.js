// @flow
import React from 'react';
import { RoundButton } from '../shared/Button';
import LabelsActionBox from '../../containers/LabelsActionBoxContainer';
import Menu from '../shared/Dropdown';
import CreateCard from '../CreateCard';
import { EditNav, Dropdown, NavContainer } from './styles';

type Props = {
    isLoading?: bool,
    onCreate?: () => void,
    showMenu?: bool,
    closeMenu?: () => void,
    collectionId?: string,
    collections?: Array<Object>,
    saveCard?: () => void,
    updateCollection?: () => void,
    showLabels?: bool,
    onShowLabels?: () => void
}

export default ({ ...props }: Props) => {
    return (
        <EditNav>
            <NavContainer>
                <RoundButton
                    text="Share post"
                    onClick={props.onCreate}
                    type="disabled"
                    isLoading={props.isLoading}
                />
                {props.showMenu &&
                    <Dropdown style={{ right: '10px' }}>
                        <Menu onClose={props.closeMenu}>
                            <CreateCard
                                collectionId={props.collectionId}
                                collections={props.collections}
                                addCard={props.saveCard}
                                updateCollection={props.updateCollection}
                                showLabels={props.showLabels}
                                onShowLabels={props.onShowLabels}
                            />
                        </Menu>
                    </Dropdown>
                }
            </NavContainer>
        </EditNav>
    );
}
