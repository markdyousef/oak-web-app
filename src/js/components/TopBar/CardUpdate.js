// @flow
import React from 'react';
import { RoundButton } from '../shared/Button';
import LabelsActionBox from '../../containers/LabelsActionBoxContainer';
import { EditNav, LabelsContainer, Labels, IconButton } from './styles';
import Dropdown from '../shared/Dropdown';
import CommentsIcon from '../../icons/comments';
import LikesIcon from '../../icons/likes';

type Props = {
    isLoading?: bool,
    onSave?: () => void,
    onShowLabels?: (show: bool) => void,
    showLabels?: bool,
    readOnly?: bool,
    onEdit?: () => void,
    onSave?: () => void,
    onShowComments?: () => void
}

export default ({ ...props }: Props) => {
    const {
        showLabels,
        onShowLabels,
        onSave,
        isLoading,
        readOnly,
        onEdit
    } = props;

    let cardActions;
    if (readOnly) {
        cardActions =
            <RoundButton
                text="Edit post"
                type="secondaryAction"
                onClick={onEdit}
                type="white"
            />
    } else {
        cardActions = [
            <Labels key="labels">
                <RoundButton
                    text="Add label"
                    onClick={() => onShowLabels && onShowLabels(!showLabels)}
                    type="white"
                />
                    {showLabels &&
                        <LabelsContainer>
                            <Dropdown onClose={() => onShowLabels && onShowLabels(false)}>
                                <LabelsActionBox />
                            </Dropdown>
                        </LabelsContainer>
                    }
            </Labels>,
            <RoundButton
                key="save"
                text="Save post"
                onClick={onSave}
                type="secondary"
                isLoading={isLoading}
            />
        ]
    }

    return (
        <EditNav>
            <IconButton>
                <span data-title="Comment" data-title-pos="bottom">
                    <CommentsIcon onClick={props.onShowComments} />
                </span>
                {<span>0</span> }
            </IconButton>
            {<IconButton>
                <span data-title="Like" data-title-pos="bottom">
                    <LikesIcon />
                </span>
                <span>0</span>
            </IconButton> }
            {cardActions}
        </EditNav>
    );
}
