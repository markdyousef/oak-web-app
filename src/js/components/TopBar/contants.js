// @flow

// ANALYTICS
export const NEW: string = 'ACTIONBAR/NEW';
export const CREATE: string = 'ACTIONBAR/CREATE';
export const SHOW_LABELS: string = 'ACTIONBAR/SHOW_LABELS';
export const SAVE: string = 'ACTIONBAR/SAVE';
export const SHOW_COMMENTS: string = 'ACTIONBAR/SHOW_COMMENTS';
export const EDIT: string = 'ACTIONBAR/EDIT';
