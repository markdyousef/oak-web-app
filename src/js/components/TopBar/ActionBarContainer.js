// @flow
import { graphql, compose } from 'react-apollo';
import gql from 'graphql-tag';
import { withRouter } from 'react-router';
import { connect } from 'react-redux';
import { labels, card, comments } from '../../store/actions';
import ActionBar from './ActionBar';
import { getTeam } from '../../utils';
import withAnalytics from './TopBarAnalytics';

const getCollectionList = gql`
    query topNav($teamId: ID!) {
        groves(teamId: $teamId) {
            id
            name
        }
    }
`;

const createCardDetail = gql`
    mutation createCardDetail($groveId: ID!, $name: String!, $content: String, $coverId: ID) {
        createSeed(groveId: $groveId, name: $name, content: $content, coverId: $coverId) {
            id
            content
        }
    }
`;

const updateCardDetail = gql`
    mutation updateCardDetail($id: ID!, $content: String!, $coverId: ID) {
        updateSeed(id: $id, content: $content, coverId: $coverId) {
            id
            content
        }
    }
`;

const mapStateToProps = (state: Object) => (
    {
        cardId: state.card.get('cardId'),
        collectionId: state.card.get('collectionId'),
        card: state.card,
        comments: state.comments,
        showLabels: state.labels.get('showLabels'),
        showComments: state.comments.get('showComments')
    }
);

type Field = {
    key: string,
    value: string
}

const mapDispatchToProps = (dispatch: Function) => (
    {
        updateLabels: (field: Field) =>
            dispatch(labels.updateLabels(field)),
        updateCard: (key: string, value: any) => {
            const field = { key, value };
            dispatch(card.updateCard(field))
        },
        updateCardContent: (content: string) =>
            dispatch(card.setCardContent(content)),
        onShowComments: (show: bool) =>
            dispatch(comments.updateComments({
                key: 'showComments',
                value: show
            }))
    }
);

export default compose(
    connect(mapStateToProps, mapDispatchToProps),
    graphql(getCollectionList, {
        options: () => ({ variables: { teamId: getTeam() } }),
        props: ({ ownProps, data: { loading, groves } }) => ({
            data: { loading, collections: groves || [] }
        })
    }),
    graphql(createCardDetail, {
        props: ({ ownProps, mutate }) => ({
            create: (groveId, name, content, coverId) =>
                mutate({ variables: { groveId, name, content, coverId } })
                    .then(res => {
                        const id = res.data.createSeed.id;
                        ownProps.updateCard('message', null);
                        ownProps.updateCard('isLoading', false);
                        ownProps.updateCard('shouldUpdate', true);
                        ownProps.updateCard('cardId', id);
                        ownProps.updateCard('isSaved', true);
                        return id;
                    })
                    .catch((err) => {
                        const message = {
                            type: 'error',
                            message: "We couldn't create your card"
                        };
                        ownProps.updateCard('message', message);
                        ownProps.updateCard('isLoading', false);
                    })
        })
    }),
    graphql(updateCardDetail, {
        props: ({ ownProps, mutate }) => ({
            update: (id:string, content:string, coverId:string) =>
                mutate({ variables: { id, content, coverId } })
                    .then(res => {
                        const { id, content } = res.data.updateSeed;
                        ownProps.updateCard('shouldUpdate', true);
                        ownProps.updateCardContent(content);
                        ownProps.updateCard('message', null);
                        ownProps.updateCard('isLoading', false);
                        ownProps.updateCard('isSaved', true);
                        return id;
                    })
                    .catch(() => {
                        const message = {
                            type: 'error',
                            message: "We couldn't update your card"
                        };
                        ownProps.updateCard('message', message);
                        ownProps.updateCard('isLoading', false);
                    })
        })
    }),
)(withRouter(withAnalytics(ActionBar)));
