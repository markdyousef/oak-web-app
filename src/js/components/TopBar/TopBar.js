// @flow
import React from 'react';
import { withRouter } from 'react-router';
import MainNavContainer from './MainNavContainer';
import ActionBarContainer from './ActionBarContainer';
import SettingsContainer from './SettingsContainer';

import { Container } from './styles';

export default () => {
    return (
        <Container>
            <MainNavContainer />
            <ActionBarContainer />
            <SettingsContainer />
        </Container>
    );
};
