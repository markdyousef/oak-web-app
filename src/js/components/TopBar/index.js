// @flow
import TopBar from './TopBar';
import MainNav from './MainNav';
import MainNavContainer from './MainNavContainer';
import Settings from './Settings';
import SettingsContainer from './SettingsContainer';
import ActionBar from './ActionBar';
import ActionBarContainer from './ActionBarContainer';
import CardUpdate from './CardUpdate';
import CardCreate from './CardCreate';

export default TopBar;

export {
    TopBar,
    MainNav,
    MainNavContainer,
    Settings,
    SettingsContainer,
    ActionBar,
    ActionBarContainer,
    CardCreate,
    CardUpdate
};
