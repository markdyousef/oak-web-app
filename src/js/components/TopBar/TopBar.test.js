// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import TopBar from './TopBar';

test('renders TopBar', () => {
    const props = {
        data: {}
    };
    const component = renderer.create(
        <TopBar {...props} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
