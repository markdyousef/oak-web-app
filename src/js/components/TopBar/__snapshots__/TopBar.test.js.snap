exports[`test renders TopBar 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <styled.nav>
    <Apollo(withRouter(MainNav)) />
    <Connect(Apollo(Apollo(Apollo(withRouter(ActionBar))))) />
    <Apollo(Apollo(withRouter(Settings))) />
</styled.nav>,
  "nodes": Array [
    <styled.nav>
      <Apollo(withRouter(MainNav)) />
      <Connect(Apollo(Apollo(Apollo(withRouter(ActionBar))))) />
      <Apollo(Apollo(withRouter(Settings))) />
</styled.nav>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Unknown />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <styled.nav>
          <Apollo(withRouter(MainNav)) />
          <Connect(Apollo(Apollo(Apollo(withRouter(ActionBar))))) />
          <Apollo(Apollo(withRouter(Settings))) />
</styled.nav>,
        "_debugID": 2,
        "_renderedOutput": <styled.nav>
          <Apollo(withRouter(MainNav)) />
          <Connect(Apollo(Apollo(Apollo(withRouter(ActionBar))))) />
          <Apollo(Apollo(withRouter(Settings))) />
</styled.nav>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Unknown />,
}
`;
